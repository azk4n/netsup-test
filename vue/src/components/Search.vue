<template>
  <div class="container-fluid">
    <div>
      <b-card
        header="Pessoas"
        header-tag="header">
        <b-alert
          variant="danger"
          dismissible
          fade
          :show="showAlertError"
          @dismissed="showAlertError=0"
        > {{errMsg}}
        </b-alert>
        <b-form inline>
          <b-input-group>
            <b-form-input id="inputId" v-model="form.id" @blur="getPessoa" placeholder="ID" />
            <b-input-group-append>
              <b-button variant="info" id="showModalButton" @click="showModal()"><font-awesome-icon icon="search" /></b-button>
            </b-input-group-append>
          </b-input-group>
          <b-input-group>
            <b-form-input v-model="form.nome" placeholder="Nome" />
          </b-input-group>
        </b-form>
      </b-card>
    </div>
    <!-- Modal Component -->
    <b-modal ref="searchModal" title="Pesquisar Pessoas">
      <b-alert
        variant="danger"
        dismissible
        fade
        :show="showModalAlertError"
        @dismissed="showModalAlertError=0"
      > {{errMsg}}
      </b-alert>
      <b-form inline>
        <b-input-group>
          <b-form-input v-model="searchForm.nome" type="text" placeholder="Nome" />
          <b-form-input v-model="searchForm.idade" type="text" placeholder="Idade" />
          <b-form-input v-model="searchForm.profissao" type="text" placeholder="ProfissÃ£o" />
          <b-button id="searchButton" @click="search()" variant="primary">Pesquisar</b-button>
        </b-input-group>
      </b-form>
      <div>
        <b-table
          id="pessoaTable"
          hover
          :items="pessoas"
          @row-dblclicked="dblClickRowHandler"
          @row-clicked="clickRowHandler"
        />
      </div>
      <div slot="modal-footer">
        <b-button :disabled="!this.selectedRow" class="float-right" variant="primary" @click="selectButtonHandler()">Selecionar</b-button>
        <b-button id="hideModalButton" class="float-right" variant="danger" @click="hideModal()">Fechar</b-button>
      </div>
    </b-modal>
  </div>
</template>

<script>
import { mapState } from 'vuex'

export default {
  name: 'Search',
  mounted () {
    this.search()
  },
  data () {
    return {
      selectedRow: '',
      showSelectButton: false,
      showAlertError: 0,
      showModalAlertError: 0,
      dismissSecsAlert: 3,
      errMsg: '',
      form: {
        id: '',
        nome: ''
      },
      searchForm: {
        nome: '',
        profissao: '',
        idade: ''
      },
      msg: ''
    }
  },
  computed: {
    ...mapState({
      pessoas: state => state.pessoas,
      pessoa: state => state.pessoa
    })
  },
  methods: {
    showModal () {
      this.$refs.searchModal.show()
    },
    hideModal () {
      this.$refs.searchModal.hide()
    },
    clickRowHandler (record, index) {
      this.showSelectButton = true
      this.selectedRow = record
    },
    dblClickRowHandler (record, index) {
      this.form = record
      this.selectedRow = ''
      this.hideModal()
    },
    selectButtonHandler () {
      this.form = this.selectedRow
      this.selectedRow = ''
      this.hideModal()
    },
    search () {
      this.$store.dispatch(
        'list',
        this.searchForm)
        .then(response => {
        })
        .catch(err => {
          this.searchForm.nome = ''
          this.showModalAlertError = 1
          this.errMsg = err.data.message
        })
    },
    getPessoa () {
      let id = this.form.id
      this.$store.dispatch(
        'detail',
        { id })
        .then((res) => {
          this.form.nome = res.result.nome
        })
        .catch(err => {
          this.form.nome = ''
          this.showAlertError = 1
          this.errMsg = err.data.message
        })
    }
  }
}
</script>
