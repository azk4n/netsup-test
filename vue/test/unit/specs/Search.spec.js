import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Search from '@/components/Search'
import Vue from 'vue'
import BootstrapVue from 'bootstrap-vue'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faSearch } from '@fortawesome/free-solid-svg-icons'

const localVue = createLocalVue()
library.add(faSearch)
localVue.use(Vuex)
localVue.use(BootstrapVue)
localVue.component('font-awesome-icon', FontAwesomeIcon)


describe('Search.vue', () => {
  let actions
  let store

  beforeEach(() => {
    // mock das acoes da store
    actions = {
      detail: jest.fn(),
      list: jest.fn()
    }
    store = new Vuex.Store({
      actions
    })
  })

  it('Ã© uma Vue instance', () => {
    const wrapper = mount(Search, { store, localVue })
    expect(wrapper.isVueInstance()).toBeTruthy()
  })
  it('renderiza corretamente', () => {
    const wrapper = mount(Search, { store, localVue })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('aciona action "detail" no @blur do input de ID', () => {
    const wrapper = mount(Search, { store, localVue })
    const input = wrapper.find('#inputId')
    input.element.value = 1
    input.trigger('blur')
    expect(actions.detail).toHaveBeenCalled()
  })

  it('aciona action "list" no @click do botao "Listar"', () => {
    const wrapper = mount(Search, { store, localVue })
    const btn = wrapper.find('#searchButton')
    btn.trigger('click')
    expect(actions.list).toHaveBeenCalled()
  })

  it('botao de "lupa" abre modal', () => {
    // usando spyOn para mockar metodo do componente
    const spy = jest.spyOn(Search.methods, 'showModal');
    const wrapper = mount(Search, { store, localVue })
    const btn = wrapper.find('#showModalButton')
    btn.trigger('click')
    expect(spy).toHaveBeenCalled()
  })

  it('botao "fechar" fecha modal', () => {
    // usando spyOn para mockar metodo do componente
    const spy = jest.spyOn(Search.methods, 'hideModal');
    const wrapper = mount(Search, { store, localVue })
    const btn = wrapper.find('#hideModalButton')
    btn.trigger('click')
    expect(spy).toHaveBeenCalled()
  })

})