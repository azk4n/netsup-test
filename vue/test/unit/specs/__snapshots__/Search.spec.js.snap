// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search.vue renderiza corretamente 1`] = `
<div
  class="container-fluid"
>
  <div>
    <div
      class="card"
    >
      <!---->
      <header
        class="card-header"
      >
        <div>
          Pessoas
        </div>
      </header>
      <div
        class="card-body"
      >
        <!---->
        <!---->
        <!---->
         
        <form
          class="form-inline"
        >
          <div
            class="input-group"
            role="group"
          >
            <!---->
            <!---->
            <input
              class="form-control"
              id="inputId"
              placeholder="ID"
              type="text"
            />
             
            <div
              class="input-group-append"
            >
              <button
                class="btn btn-info"
                id="showModalButton"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-search fa-w-16"
                  data-icon="search"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class=""
                    d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
            <!---->
            <!---->
          </div>
           
          <div
            class="input-group"
            role="group"
          >
            <!---->
            <!---->
            <input
              class="form-control"
              placeholder="Nome"
              type="text"
            />
            <!---->
            <!---->
          </div>
        </form>
      </div>
      <!---->
      <!---->
    </div>
  </div>
   
  <div>
    <div
      style="position: relative; z-index: 2000;"
    >
      <div
        aria-hidden="true"
        class="modal fade"
        role="dialog"
        style="display: none;"
        tabindex="-1"
      >
        <div
          class="modal-dialog modal-md"
        >
          <div
            class="modal-content"
            role="document"
          >
            <header
              class="modal-header"
            >
              <h5
                class="modal-title"
              >
                Pesquisar Pessoas
              </h5>
              <button
                aria-label="Close"
                class="close"
                type="button"
              >
                ×
              </button>
            </header>
            <div
              class="modal-body"
            >
              <!---->
               
              <form
                class="form-inline"
              >
                <div
                  class="input-group"
                  role="group"
                >
                  <!---->
                  <!---->
                  <input
                    class="form-control"
                    placeholder="Nome"
                    type="text"
                  />
                   
                  <input
                    class="form-control"
                    placeholder="Idade"
                    type="text"
                  />
                   
                  <input
                    class="form-control"
                    placeholder="Profissão"
                    type="text"
                  />
                   
                  <button
                    class="btn btn-primary"
                    id="searchButton"
                    type="button"
                  >
                    Pesquisar
                  </button>
                  <!---->
                  <!---->
                </div>
              </form>
               
              <div>
                <table
                  aria-busy="false"
                  aria-colcount="0"
                  class="table b-table table-hover"
                  id="pessoaTable"
                >
                  <!---->
                  <!---->
                  <thead
                    class=""
                  >
                    <tr />
                  </thead>
                  <!---->
                  <tbody
                    class=""
                  >
                    <!---->
                    <!---->
                    <!---->
                  </tbody>
                </table>
              </div>
               
            </div>
            <footer
              class="modal-footer"
            >
              <div>
                <button
                  class="btn float-right btn-primary disabled"
                  disabled="disabled"
                  type="button"
                >
                  Selecionar
                </button>
                 
                <button
                  class="btn float-right btn-danger"
                  id="hideModalButton"
                  type="button"
                >
                  Fechar
                </button>
              </div>
            </footer>
          </div>
        </div>
      </div>
      <!---->
      <!---->
    </div>
  </div>
</div>
`;
